# Task 1.3------Question 1-----Descriptive Statistics

### Numerical Descriptive Statistics:

In [110]:
bank['age'].describe()


count    37084.000000
mean        40.042714
std         10.432965
min         17.000000
25%         32.000000
50%         38.000000
75%         47.000000
max         98.000000
Name: age, dtype: float64

In [120]:
bank['emp.var.rate'].describe()

count    37084.000000
mean         0.082669
std          1.568997
min         -3.400000
25%         -1.800000
50%          1.100000
75%          1.400000
max          1.400000
Name: emp.var.rate, dtype: float64

In [143]:
bank['cons.price.idx'].describe()

count    37084.000000
mean        93.576076
std          0.578493
min         92.201000
25%         93.075000
50%         93.749000
75%         93.994000
max         94.767000
Name: cons.price.idx, dtype: float64

In [144]:
bank['cons.conf.idx'].describe()

count    37084.000000
mean       -40.505183
std          4.622045
min        -50.800000
25%        -42.700000
50%        -41.800000
75%        -36.400000
max        -26.900000
Name: cons.conf.idx, dtype: float64

In [145]:
bank['euribor3m'].describe()

count    37084.000000
mean         3.621668
std          1.733972
min          0.634000
25%          1.344000
50%          4.857000
75%          4.961000
max          5.045000
Name: euribor3m, dtype: float64

In [146]:
bank['nr.employed'].describe()

count    37084.000000
mean      5167.058664
std         72.196605
min       4963.600000
25%       5099.100000
50%       5191.000000
75%       5228.100000
max       5228.100000
Name: nr.employed, dtype: float64

## Age:

In [88]:
print(bank['age'].unique())
print(bank['age'].value_counts())

[56 45 59 41 24 25 29 57 35 54 46 50 39 55 37 49 34 52 58 32 38 44 42 40
 60 53 47 51 48 33 31 43 36 28 30 27 22 26 23 20 21 61 19 18 70 66 76 67
 73 88 95 77 68 75 80 62 63 65 72 82 64 71 69 78 85 79 83 81 74 17 91 86
 98 94 84 92 89]
31    1761
33    1656
32    1644
36    1627
35    1584
      ... 
98       2
89       2
91       2
94       1
95       1
Name: age, Length: 77, dtype: int64


In [123]:
print('mean age group of clients not subscribed',bank[bank['y']=='no']['age'].mean())
print('mean age group of clients subscribed',bank[bank['y']=='yes']['age'].mean())
print('median age group of clients not subscribed',bank[bank['y']=='no']['age'].median())
print('median age group of clients subscribed',bank[bank['y']=='yes']['age'].median())
print('range of  age:',bank['age'].min(),"to",bank['age'].max())
print("mode:",bank[bank['y']=='yes']['age'].mode())


mean age group of clients not subscribed 39.92493004015087
mean age group of clients subscribed 40.96292775665399
median age group of clients not subscribed 38.0
median age group of clients subscribed 37.0
range of  age: 17 to 98
mode: 0    31
dtype: int64


In [77]:
age,k=bank[['age','y']].copy(deep=True),[]
for i in bank['age']:   #dividing the age into groups
    if i<26:
        k.append('Group<25')
    elif i>25 and i<=45:
        k.append('25<Group<=45')
    elif i>45 and i<=65:
        k.append('45<Group<=65')
    elif i>65:
        k.append('Group>65')
    else:
        continue
age['Groups']=k

df=pd.crosstab(age['Groups'],age['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df[['total','yes %','no %']])

y             total      yes %       no %
Groups                                   
25<Group<=45  24912  10.280186  89.719814
45<Group<=65  10114  10.529958  89.470042
Group<25       1492  21.045576  78.954424
Group>65        566  47.349823  52.650177


# Job:

In [83]:
print(bank['job'].unique())
print(bank['job'].value_counts())

['services' 'admin.' 'blue-collar' 'technician' 'housemaid' 'retired'
 'management' 'unknown' 'entrepreneur' 'unemployed' 'student'
 'self-employed']
admin.           9420
blue-collar      8314
technician       6096
services         3547
management       2637
retired          1541
entrepreneur     1309
self-employed    1276
housemaid         946
unemployed        925
student           767
unknown           306
Name: job, dtype: int64


In [131]:
print('count of clients not subscribed:\n',bank[bank['y']=='no']['job'].value_counts())
print('count of clients  subscribed:\n',bank[bank['y']=='yes']['job'].value_counts())

count of clients not subscribed:
 admin.           8174
blue-collar      7738
technician       5439
services         3249
management       2345
entrepreneur     1198
retired          1146
self-employed    1145
housemaid         853
unemployed        791
student           527
unknown           271
Name: job, dtype: int64
count of clients  subscribed:
 admin.           1246
technician        657
blue-collar       576
retired           395
services          298
management        292
student           240
unemployed        134
self-employed     131
entrepreneur      111
housemaid          93
unknown            35
Name: job, dtype: int64


In [79]:
print('Percentage of distribution of the data\n')
print(np.round(bank['job'].value_counts()/len(bank)*100,2))

Percentage of distribution of the data

admin.           25.40
blue-collar      22.42
technician       16.44
services          9.56
management        7.11
retired           4.16
entrepreneur      3.53
self-employed     3.44
housemaid         2.55
unemployed        2.49
student           2.07
unknown           0.83
Name: job, dtype: float64


In [80]:
df=pd.crosstab(bank['job'],bank['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df[['total','yes %','no %']])

y              total      yes %       no %
job                                       
admin.          9420  13.227176  86.772824
blue-collar     8314   6.928073  93.071927
entrepreneur    1309   8.479756  91.520244
housemaid        946   9.830867  90.169133
management      2637  11.073189  88.926811
retired         1541  25.632706  74.367294
self-employed   1276  10.266458  89.733542
services        3547   8.401466  91.598534
student          767  31.290743  68.709257
technician      6096  10.777559  89.222441
unemployed       925  14.486486  85.513514
unknown          306  11.437908  88.562092


In [135]:
yes=pd.DataFrame(
    np.round(bank[bank['y']=='yes']['job'].value_counts()/len(bank['job'])*100,2)).reset_index()
no=pd.DataFrame(
    np.round(bank[bank['y']=='no']['job'].value_counts()/len(bank['job'])*100,2)).reset_index()
distribution=pd.merge(yes,no,on='index').rename(columns={'index':'jobs',  'job_x':'subscribed%' , 
                                                         'job_y':'not subscribed%'})
print('distribution of data ')
print(distribution)

distribution of data 
             jobs  subscribed%  not subscribed%
0          admin.         3.36            22.04
1      technician         1.77            14.67
2     blue-collar         1.55            20.87
3         retired         1.07             3.09
4        services         0.80             8.76
5      management         0.79             6.32
6         student         0.65             1.42
7      unemployed         0.36             2.13
8   self-employed         0.35             3.09
9    entrepreneur         0.30             3.23
10      housemaid         0.25             2.30
11        unknown         0.09             0.73


# Marital:

In [23]:
df=pd.crosstab(bank['marital'],bank['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

y            no   yes  total       no %      yes %
marital                                           
divorced   3699   427   4126  89.650994  10.349006
married   20187  2292  22479  89.803817  10.196183
single     8930  1477  10407  85.807629  14.192371
unknown      60    12     72  83.333333  16.666667


In [27]:
print('count of clients not subscribed:\n',bank[bank['y']=='no']['marital'].value_counts())
print('count of clients  subscribed:\n',bank[bank['y']=='yes']['marital'].value_counts())

count of clients not subscribed:
 married     20187
single       8930
divorced     3699
unknown        60
Name: marital, dtype: int64
count of clients  subscribed:
 married     2292
single      1477
divorced     427
unknown       12
Name: marital, dtype: int64


In [24]:

yes=pd.DataFrame(bank[bank['y']=='yes']['marital'].value_counts()/len(bank['marital'])*100).reset_index()
no=pd.DataFrame(bank[bank['y']=='no']['marital'].value_counts()/len(bank['marital'])*100).reset_index()

distribution=pd.merge(yes,no,on='index').rename(columns={'index':'marital','marital_x':'subscribed%' ,
                                                         'marital_y':'not subscribed%'})
print('distribution of data ')
print(distribution)


distribution of data 
    marital  subscribed%  not subscribed%
0   married     6.180563        54.435875
1    single     3.982850        24.080466
2  divorced     1.151440         9.974652
3   unknown     0.032359         0.161795


# Education:

In [25]:
df=pd.crosstab(bank['education'],bank['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

y                      no   yes  total       no %      yes %
education                                                   
basic.4y             3374   391   3765  89.614874  10.385126
basic.6y             1903   171   2074  91.755063   8.244937
basic.9y             5001   420   5421  92.252352   7.747648
high.school          7597   945   8542  88.937017  11.062983
illiterate             14     2     16  87.500000  12.500000
professional.course  4205   541   4746  88.600927  11.399073
university.degree    9464  1507  10971  86.263786  13.736214
unknown              1318   231   1549  85.087153  14.912847


In [28]:
print('count of clients not subscribed:\n',bank[bank['y']=='no']['education'].value_counts())
print('count of clients  subscribed:\n',bank[bank['y']=='yes']['education'].value_counts())

count of clients not subscribed:
 university.degree      9464
high.school            7597
basic.9y               5001
professional.course    4205
basic.4y               3374
basic.6y               1903
unknown                1318
illiterate               14
Name: education, dtype: int64
count of clients  subscribed:
 university.degree      1507
high.school             945
professional.course     541
basic.9y                420
basic.4y                391
unknown                 231
basic.6y                171
illiterate                2
Name: education, dtype: int64


In [26]:
yes=pd.DataFrame(bank[bank['y']=='yes']['education'].value_counts()/len(bank['education'])*100).reset_index()
no=pd.DataFrame(bank[bank['y']=='no']['education'].value_counts()/len(bank['education'])*100).reset_index()

distribution=pd.merge(yes,no,on='index').rename(columns={'index':'education','education_x':'subscribed%' ,
                                                         'education_y':'not subscribed%'})
print('Distribution of data: \n')
print(distribution)

print('education proportions of the subscribers:\n')
print(bank[bank['y']=='yes']['education'].value_counts()/len(bank[bank['y']=='yes']['education'])*100)
print('education proportions of the non-subscribers:\n')
print(bank[bank['y']=='no']['education'].value_counts()/len(bank[bank['y']=='no']['education'])*100)

Distribution of data: 

             education  subscribed%  not subscribed%
0    university.degree     4.063747        25.520440
1          high.school     2.548269        20.485924
2  professional.course     1.458850        11.339122
3             basic.9y     1.132564        13.485600
4             basic.4y     1.054363         9.098263
5              unknown     0.622910         3.554093
6             basic.6y     0.461115         5.131593
7           illiterate     0.005393         0.037752
education proportions of the subscribers:

university.degree      35.812738
high.school            22.457224
professional.course    12.856464
basic.9y                9.980989
basic.4y                9.291825
unknown                 5.489544
basic.6y                4.063688
illiterate              0.047529
Name: education, dtype: float64
education proportions of the non-subscribers:

university.degree      28.786957
high.school            23.108042
basic.9y               15.211705
professional.c

# Default: 

Have Credit in default? (Yes or No)

In [29]:
df=pd.crosstab(bank['default'],bank['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

y           no   yes  total        no %      yes %
default                                           
no       25576  3806  29382   87.046491  12.953509
unknown   7298   402   7700   94.779221   5.220779
yes          2     0      2  100.000000   0.000000


In [39]:
print('count of clients not subscribed:\n',bank[bank['y']=='no']['default'].value_counts())
print('count of clients  subscribed:\n',bank[bank['y']=='yes']['default'].value_counts())

count of clients not subscribed:
 no         25576
unknown     7298
yes            2
Name: default, dtype: int64
count of clients  subscribed:
 no         3806
unknown     402
Name: default, dtype: int64


In [31]:
print('percentage of subscriber has credit card  and doesnot has credit card')
print(bank[bank['y']=='yes']['default'].value_counts()/len(bank[bank['y']=='yes']['default'])*100)


percentage of subscriber has credit card  and doesnot has credit card
no         90.446768
unknown     9.553232
Name: default, dtype: float64


99 % customer doesnot has credit card who has subscribed.

In [32]:
print('percentage of non subscriber has credit card  and doesnot has credit card')

print(bank[bank['y']=='no']['default'].value_counts()/len(bank[bank['y']=='no']['default'])*100)

percentage of non subscriber has credit card  and doesnot has credit card
no         77.795352
unknown    22.198564
yes         0.006083
Name: default, dtype: float64


# Housing:

has housing loan? (binary: "yes","no")

In [33]:
df=pd.crosstab(bank['housing'],bank['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

y           no   yes  total       no %      yes %
housing                                          
no       14931  1838  16769  89.039299  10.960701
unknown    791    91    882  89.682540  10.317460
yes      17154  2279  19433  88.272526  11.727474


In [41]:
print('count of clients not subscribed:\n',bank[bank['y']=='no']['housing'].value_counts())
print('count of clients  subscribed:\n',bank[bank['y']=='yes']['housing'].value_counts())

count of clients not subscribed:
 yes        17154
no         14931
unknown      791
Name: housing, dtype: int64
count of clients  subscribed:
 yes        2279
no         1838
unknown      91
Name: housing, dtype: int64


In [34]:
print('percentage of non subscribers v/s have/dont have housing loan')
print(bank[bank['y']=='no']['housing'].value_counts()/len(bank[bank['y']=='no']['housing'])*100)
print('percentage of  subscribers v/s have/dont have housing loan')
print(bank[bank['y']=='yes']['housing'].value_counts()/len(bank[bank['y']=='yes']['housing'])*100)

percentage of non subscribers v/s have/dont have housing loan
yes        52.177881
no         45.416109
unknown     2.406010
Name: housing, dtype: float64
percentage of  subscribers v/s have/dont have housing loan
yes        54.158745
no         43.678707
unknown     2.162548
Name: housing, dtype: float64


#### conclusion:
    58 % non subscribers has housing loan
    63 % subscribers dont have housing loan

In [42]:
yes=pd.DataFrame(bank[bank['y']=='yes']['housing'].value_counts()/len(bank['housing'])*100).reset_index()
no=pd.DataFrame(bank[bank['y']=='no']['housing'].value_counts()/len(bank['housing'])*100).reset_index()

distribution=pd.merge(yes,no,on='index').rename(columns={'index':'housingLoan','housing_x':'subscribed%' ,
                                                         'housing_y':'not subscribed%'})
print('percentage of the client who have/donthave housing Loan')
print(distribution)

percentage of the client who have/donthave housing Loan
  housingLoan  subscribed%  not subscribed%
0         yes     6.145507        46.257146
1          no     4.956315        40.262647
2     unknown     0.245389         2.132995


# Loan:

Loan: has personal loan? (binary: "yes","no")


   Note: 83 percent of the clients donot have personal loan

In [44]:
yes=pd.DataFrame(bank[bank['y']=='yes']['loan'].value_counts()/len(bank['loan'])*100).reset_index()
no=pd.DataFrame(bank[bank['y']=='no']['loan'].value_counts()/len(bank['loan'])*100).reset_index()

distribution=pd.merge(yes,no,on='index').rename(columns={'index':'PersonalLoan','loan_x':'subscribed%' ,
                                                         'loan_y':'not subscribed%'})
print('percentage of the client who has/hasnot Personal Loan')
print(distribution)

percentage of the client who has/hasnot Personal Loan
  PersonalLoan  subscribed%  not subscribed%
0           no     9.435336        72.974868
1          yes     1.666487        13.544925
2      unknown     0.245389         2.132995


In [47]:
print('count of clients not subscribed:\n',bank[bank['y']=='no']['loan'].value_counts())
print('count of clients  subscribed:\n',bank[bank['y']=='yes']['loan'].value_counts())

count of clients not subscribed:
 no         27062
yes         5023
unknown      791
Name: loan, dtype: int64
count of clients  subscribed:
 no         3499
yes         618
unknown      91
Name: loan, dtype: int64


In [45]:
df=pd.crosstab(bank['loan'],bank['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

y           no   yes  total       no %      yes %
loan                                             
no       27062  3499  30561  88.550767  11.449233
unknown    791    91    882  89.682540  10.317460
yes       5023   618   5641  89.044496  10.955504


In [46]:
print('proportion of subscribers have personal loan')
print(bank[bank['y']=='yes']['loan'].value_counts()/len(bank[bank['y']=='yes']['loan'])*100)
print('proportion of non subscribers have personal loan')
print(bank[bank['y']=='no']['loan'].value_counts()/len(bank[bank['y']=='no']['loan'])*100)

proportion of subscribers have personal loan
no         83.151141
yes        14.686312
unknown     2.162548
Name: loan, dtype: float64
proportion of non subscribers have personal loan
no         82.315367
yes        15.278623
unknown     2.406010
Name: loan, dtype: float64


9 % of the clients who has subscribed has personal loan
16.9 % of the clients who has not subscribed has personal loan

# Last Campaign Data:

In [48]:
bank['contact'].value_counts()/len(bank['contact'])*100

cellular     63.428972
telephone    36.571028
Name: contact, dtype: float64

 Note: 63.4 % of the clients were approached through cellularcontact  and 
 36.5% of the clients were approached through telephonic communication

In [49]:
df=pd.crosstab(bank['contact'],bank['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

y             no   yes  total       no %      yes %
contact                                            
cellular   20030  3492  23522  85.154324  14.845676
telephone  12846   716  13562  94.720543   5.279457


In [50]:
yes=pd.DataFrame(bank[bank['y']=='yes']['contact'].value_counts()/len(bank['contact'])*100).reset_index()
no=pd.DataFrame(bank[bank['y']=='no']['contact'].value_counts()/len(bank['contact'])*100).reset_index()
distribution=pd.merge(yes,no,on='index').rename(columns={'index':'contact','contact_x':'subscribed%' ,
                                                         'contact_y':'not subscribed%'})
print('percentage of the client wrt Mode of Contact')
print(distribution)

percentage of the client wrt Mode of Contact
     contact  subscribed%  not subscribed%
0   cellular     9.416460        54.012512
1  telephone     1.930752        34.640276


In [51]:
print(bank[bank['y']=='no']['contact'].value_counts()/len(bank[bank['y']=='no']['contact'])*100)

cellular     60.925903
telephone    39.074097
Name: contact, dtype: float64


In [52]:
print(bank[bank['y']=='yes']['contact'].value_counts()/len(bank[bank['y']=='yes']['contact'])*100)

cellular     82.984791
telephone    17.015209
Name: contact, dtype: float64


of the people who had subscribed 82.9% clients were approached through cellular contact

# Month: 
#### Month: Last contact month:

In [58]:
(bank['day_of_week'].value_counts())

thu    7778
mon    7671
wed    7341
tue    7262
fri    7032
Name: day_of_week, dtype: int64

In [59]:

df=pd.crosstab(bank['month'],bank['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

y         no  yes  total       no %      yes %
month                                         
apr     1892  496   2388  79.229481  20.770519
aug     4973  589   5562  89.410284  10.589716
dec       84   82    166  50.602410  49.397590
jul     5839  591   6430  90.808709   9.191291
jun     4278  508   4786  89.385708  10.614292
mar      241  246    487  49.486653  50.513347
may    11607  813  12420  93.454106   6.545894
nov     3328  377   3705  89.824561  10.175439
oct      351  279    630  55.714286  44.285714
sep      283  227    510  55.490196  44.509804


In [60]:
yes=pd.DataFrame(bank[bank['y']=='yes']['month'].value_counts()/len(bank['month'])*100).reset_index()
no=pd.DataFrame(bank[bank['y']=='no']['month'].value_counts()/len(bank['month'])*100).reset_index()
distribution=pd.merge(yes,no,on='index').rename(columns={'index':'month','month_x':'subscribed%' ,
                                                         'month_y':'not subscribed%'})
print('percentage of the client wrt Mode of Contact')
print(distribution)
yes=pd.DataFrame(bank[bank['y']=='yes']['month'].value_counts()/len(bank[bank['y']=='yes']['month'])*100).reset_index()
no=pd.DataFrame(bank[bank['y']=='no']['month'].value_counts()/len(bank[bank['y']=='no']['month'])*100).reset_index()
distribution=pd.merge(yes,no,on='index').rename(columns={'index':'month','month_x':'subscribed%' ,
                                                         'month_y':'not subscribed%'})
print('percentage of the subscribers wrt Mode of Contact')
print(distribution)

percentage of the client wrt Mode of Contact
  month  subscribed%  not subscribed%
0   may     2.192320        31.299213
1   jul     1.593679        15.745335
2   aug     1.588286        13.410096
3   jun     1.369863        11.535972
4   apr     1.337504         5.101931
5   nov     1.016611         8.974221
6   oct     0.752346         0.946500
7   mar     0.663359         0.649876
8   sep     0.612124         0.763132
9   dec     0.221120         0.226513
percentage of the subscribers wrt Mode of Contact
  month  subscribed%  not subscribed%
0   may    19.320342        35.305390
1   jul    14.044677        17.760676
2   aug    13.997148        15.126536
3   jun    12.072243        13.012532
4   apr    11.787072         5.754958
5   nov     8.959125        10.122886
6   oct     6.630228         1.067648
7   mar     5.846008         0.733058
8   sep     5.394487         0.860810
9   dec     1.948669         0.255506


# Duration:

duration: last contact duration, in seconds

#### Converting into minutes:

In [61]:
minutes=bank[['duration']]/60 

In [63]:
k=[]
for i in minutes['duration']:
    if i<=5:
        k.append('<=5')
    elif i>5 and i<=10:
        k.append('>5&<=10')
    elif i>10 and i<=15:
        k.append('>10&<=15')
    elif i>15 and i<=20:
        k.append('>15&<=20')
    else:
        k.append('>20')
minutes['time']=k

In [64]:
minutes['y']=bank['y']

In [65]:
df=pd.crosstab(minutes['time'],minutes['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

y            no   yes  total       no %      yes %
time                                              
<=5       25627  1361  26988  94.957018   5.042982
>10&<=15   1142   842   1984  57.560484  42.439516
>15&<=20    294   400    694  42.363112  57.636888
>20         171   289    460  37.173913  62.826087
>5&<=10    5642  1316   6958  81.086519  18.913481


In [66]:
yes=pd.DataFrame(minutes[minutes['y']=='yes']['time'].value_counts()/len(minutes['time'])*100).reset_index()
no=pd.DataFrame(minutes[minutes['y']=='no']['time'].value_counts()/len(minutes['time'])*100).reset_index()
distribution=pd.merge(yes,no,on='index').rename(columns={'index':'time','time_x':'subscribed%' ,
                                                         'time_y':'not subscribed%'})
print('percentage of the client wrt Mode of Contact')
print(distribution)

yes=pd.DataFrame(minutes[minutes['y']=='yes']['time'].value_counts()/len(minutes[minutes['y']=='yes'])*100).reset_index()
no=pd.DataFrame(minutes[minutes['y']=='no']['time'].value_counts()/len(minutes[minutes['y']=='no'])*100).reset_index()
distribution=pd.merge(yes,no,on='index').rename(columns={'index':'time','time_x':'subscribed%' ,
                                                         'time_y':'not subscribed%'})
print('\npercentage of the client wrt Mode of Contact')
print(distribution)

percentage of the client wrt Mode of Contact
       time  subscribed%  not subscribed%
0       <=5     3.670046        69.105275
1   >5&<=10     3.548700        15.214109
2  >10&<=15     2.270521         3.079495
3  >15&<=20     1.078632         0.792795
4       >20     0.779312         0.461115

percentage of the client wrt Mode of Contact
       time  subscribed%  not subscribed%
0       <=5    32.343156        77.950481
1   >5&<=10    31.273764        17.161455
2  >10&<=15    20.009506         3.473659
3  >15&<=20     9.505703         0.894269
4       >20     6.867871         0.520136


# Campaign:

campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)

In [69]:
total=pd.DataFrame(bank['campaign'].value_counts()).reset_index().rename(
    columns={'index':'number_of_approaches','campaign':'total'})

yes=pd.DataFrame(bank[bank['y']=='yes']['campaign'].value_counts()).reset_index().rename(
    columns={'index':'number_of_approaches','campaign':'yes'})

no=pd.DataFrame(bank[bank['y']=='no']['campaign'].value_counts()).reset_index().rename(
    columns={'index':'number_of_approaches','campaign':'no'})

success=pd.merge(pd.merge(total,yes,on='number_of_approaches',how='outer'),no,on='number_of_approaches',how='outer')
success['successrate']=success[success['number_of_approaches']<=32]['yes']/success['total']*100
success

number_of_approaches  total     yes     no  successrate
0                      1  15883  2090.0  13793    13.158723
1                      2   9532  1094.0   8438    11.477130
2                      3   4777   515.0   4262    10.780825
3                      4   2389   235.0   2154     9.836752
4                      5   1441   110.0   1331     7.633588
5                      6    873    69.0    804     7.903780
6                      7    580    33.0    547     5.689655
7                      8    360    14.0    346     3.888889
8                      9    260    13.0    247     5.000000
9                     10    196    10.0    186     5.102041
10                    11    165    11.0    154     6.666667
11                    12    118     3.0    115     2.542373
12                    13     88     3.0     85     3.409091
13                    14     62     1.0     61     1.612903
14                    17     49     4.0     45     8.163265
15                    16     45     NaN     45          NaN
16                    15     41     2.0     39     4.878049
17                    18     32     NaN     32          NaN
18                    20     28     NaN     28          NaN
19                    19     25     NaN     25          NaN
20                    21     24     NaN     24          NaN
21                    24     14     NaN     14          NaN
22                    23     14     1.0     13     7.142857
23                    22     14     NaN     14          NaN
24                    27      9     NaN      9          NaN
25                    28      8     NaN      8          NaN
26                    29      8     NaN      8          NaN
27                    26      7     NaN      7          NaN
28                    25      6     NaN      6          NaN
29                    30      6     NaN      6          NaN
30                    31      6     NaN      6          NaN
31                    35      4     NaN      4          NaN
32                    33      4     NaN      4          NaN
33                    32      4     NaN      4          NaN
34                    34      3     NaN      3          NaN
35                    43      2     NaN      2          NaN
36                    40      2     NaN      2          NaN
37                    42      1     NaN      1          NaN
38                    41      1     NaN      1          NaN
39                    56      1     NaN      1          NaN
40                    39      1     NaN      1          NaN
41                    37      1     NaN      1          NaN

we see that 87% of subscribers comes from first 3 attemts only and the no of subscribers decreases from so on
such that .55% of subscribers comes from more than 10 attempts 
if we look at the success rate it is more than 10% for first 10 attempts

In [70]:
print('percentage of subscribers in first 10 attepts')
print((success[success['number_of_approaches']<=32]['yes']/success['yes'].sum()*100)[0:10])


percentage of subscribers in first 10 attepts
0    49.667300
1    25.998099
2    12.238593
3     5.584601
4     2.614068
5     1.639734
6     0.784221
7     0.332700
8     0.308935
9     0.237643
Name: yes, dtype: float64


In [71]:
(success[success['number_of_approaches']<=32]['no']/success['no'].sum()*100)[0:10]

0    41.954617
1    25.666139
2    12.963864
3     6.551892
4     4.048546
5     2.445553
6     1.663828
7     1.052439
8     0.751308
9     0.565762
Name: no, dtype: float64

In [72]:
success['no'].sum()+success['yes'].sum()

37084.0

# Pdays:
##### pdays:
number of days that passed by after the client was last contacted from a previous campaign (numeric, -1 means client was not previously contacted)


number of days that passed by after the client was last contacted from a previous campaign (numeric, 999 means client was not previously contacted)

In [74]:
k=[]
for i in bank['pdays']:
    if i==999:
        k.append('newClients')
    else:
        k.append('oldClients')
df=bank[['y']].copy(deep=True)
df['type']=k

In [75]:

df=pd.crosstab(df['type'],df['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

y              no   yes  total       no %      yes %
type                                                
newClients  32391  3331  35722  90.675214   9.324786
oldClients    485   877   1362  35.609398  64.390602


If the old client is too old >200 days then chances of non subscribing are more

# Previous:
##### previous:
number of contacts performed before this campaign and for this client (numeric)

In [79]:
print('number of newly approached clients')
print(bank[bank['pdays']==999].shape[0])
print('number of clients who were approached zero times before this')
print(bank[bank['previous']==0].shape[0])
print('number of old  clients')
print(bank[bank['pdays']<999].shape[0])
print('number of clients who were approached zero times before this')
print(bank[bank['previous']>0].shape[0])

number of newly approached clients
35722
number of clients who were approached zero times before this
32023
number of old  clients
1362
number of clients who were approached zero times before this
5061


In [80]:
k=[]
for i in bank['pdays']:
    if i==999:
        k.append('newClients')
    else:
        k.append('oldClients')
df=bank[['y']].copy(deep=True)
df['type']=k

print('proportion of old clients new clients subscribed/not subscribed')
df=pd.crosstab(df['type'],minutes['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

proportion of old clients new clients subscribed/not subscribed
y              no   yes  total       no %      yes %
type                                                
newClients  32391  3331  35722  90.675214   9.324786
oldClients    485   877   1362  35.609398  64.390602


we can say presence of high collinearity with pdays since 999 in  pday means new client which is 32391 .

# Poutcome:
poutcome: outcome of the previous marketing campaign (categorical: "unknown","other","failure","success")

In [81]:
bank['poutcome'].value_counts()

nonexistent    32023
failure         3822
success         1239
Name: poutcome, dtype: int64

In [82]:
df=pd.crosstab(bank['poutcome'],bank['y'])
df['total']=df['no']+df['yes']
df['no %']=df['no']/df['total']*100
df['yes %']=df['yes']/df['total']*100
print(df)

y               no   yes  total       no %      yes %
poutcome                                             
failure       3274   548   3822  85.661957  14.338043
nonexistent  29178  2845  32023  91.115761   8.884239
success        424   815   1239  34.221146  65.778854


# Hypothesis Testing : Task 1.3-----Question 2: